Make image (layer) downloads faster by using pigz

sargun opened the pull request:
**- What I did**
The Golang built-in gzip library is serialized, and fairly slow
at decompressing. It also only decompresses on demand, versus
pipelining decompression. This fixes those problems.

**- How I did it**
This change switches to trying to use parallel gzip for gzip decompression as opposed to the 
golang one. If it isn't available, it'll fall back to the default. This code path can also be disabled by environment variable.

**- How to verify it**
There are existing tests for this codepath to ensure correctness of this implementation. I ran some manual benchmarks, and I found I was able to get about 50% better performance as opposed to the build in library. These performance gains were primarily seen on images with layers about 10MB.

**- Description for the changelog**
Make image (layer) downloads faster by using pgzip

sargun commented:
Any idea what:
```
00:02:07 2017/12/05 00:02:07 unrecognized import path "cloud.google.com/go" (parse https://cloud.google.com/go?go-get=1: no go-import meta tags)
```
Is due to?

thaJeztah commented:
looks like this is implementing the same as https://github.com/moby/moby/pull/34788 (looks like that one stalled though)

sargun commented:
Should I hate this behind an environment variable, so you can opt out of
the library? The primary downside I know is that there is a memory overhead
with this library.

Sent from my iPhone

On Dec 4, 2017, at 21:00, Sebastiaan van Stijn <notifications@github.com>
wrote:

looks like this is implementing the same as #34788
<https://github.com/moby/moby/pull/34788> (looks like that one stalled
though)

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/moby/moby/pull/35697#issuecomment-349196097>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AAtyRNDunSsbTahMlr9qTJ9MRJN9MYsGks5s9M4LgaJpZM4Q1g4j>
.

thaJeztah commented:
We preferred to use the external binary instead of using the Go implementation (see the other pull request); unless there's compelling reasons that you know of.

sargun commented:
@thaJeztah Switched out. It looks like LZMA / xz was already doing this, so I took this opportunity to share code between the two of them.

sargun commented:
Once this is in, I'll add pigz as a recommendation to the packages here: https://github.com/docker/docker-ce-packaging.

sargun commented:
@thaJeztah I don't see an easy way to install pigz in the container for CI. Any ideas?

sargun commented:
@thaJeztah PTAL.

sargun commented:
CC: 
@ripcurld0 
@jboero

thaJeztah commented on pkg/archive/archive.go:
```diff
@@ -136,10 +138,37 @@ func DetectCompression(source []byte) Compression {
+
+	unpigzPath, err := exec.LookPath("unpigz")
+	if err != nil {
+		logrus.Debug("Cannot find unpigz")
```
Possibly change the error to be something like "Unpigz binary not found in PATH, falling back to regular gzip"  (better suggestions welcome 😅 )

thaJeztah commented:
Can you add the package to the Dockerfiles, so that it can be run in CI? (Or is it already there)?

thaJeztah commented:
ping @AkihiroSuda @unclejack @stevvooe PTAL

AkihiroSuda commented:
I don't think we should add an environment variable.

I suggest introducing a daemon flag `--parallel-decompression=pigz|none` and pass it to pkg/archive via some `WithPigz DecompressOpt` like
```go
func DecompressStream(archive io.Reader, opts DecompressOpt...) (io.ReadCloser, error)
```

sargun commented:
@AkihiroSuda Exposing that knob seems like overkill, and introduces a deprecation cycle if we ever want to revert to using the Golang parallel gzip library.

sargun commented:
@AkihiroSuda If we do anything, I'd suggest (part of a different PR), `--decompress-with=gzip|unpigz|internal`

AkihiroSuda commented:
> --decompress-gz-with=gzip|unpigz|internal

What is gzip here? Do you mean executing /bin/gzip?

sargun commented:
@AkihiroSuda correct.

AkihiroSuda commented:
How /bin/gzip is useful?

sargun commented:
-- But, I think making this (obviously) behaviour configurable opens up a configuration troubleshooting nightmare if anyone changes it.

sargun commented:
/bin/gzip is faster than internal gzip.

stevvooe commented:
This has already been discussed in detail. Last time we went through this exercise, a knob was considered necessary.

Note that this risks breaking _everyone's_ image cache, so this is not a light consideration.

I'd prefer that we allow the daemon to disable layer compression altogether and move towards using transport compression. This would allow us to choose the fastest, rather than be stuck on gzip, which is wildly slow.

AkihiroSuda commented:
> Note that this risks breaking everyone's image cache, so this is not a light consideration.

How decompression breaks image cache?

sargun commented:
@stevvooe I'm unsure how this breaks image caching. There is a knob, it's just a "hidden" knob (environment variable). If it turns out this is breaking everyone, people will know what the environment variable is, otherwise there is no reason to disable this feature.

sargun commented:
@stevvooe I can do another PR (once my current two clear), that will avoid compression of images on docker push. I think it's more complicated because it involves some sort of optional in the registry and the registry to understand tar+compression.

sargun commented:
@stevvooe Bump?

thaJeztah commented:
Discussing in the maintainers meeting and we are okay with these changes (also per https://github.com/moby/moby/pull/34788#issuecomment-341515005). Moving this to code review.


@tonistiigi had a note about the `cmdstream` but I leave that for him to comment :smile:

stevvooe commented on pkg/archive/archive_test.go:
```diff
@@ -1318,3 +1320,74 @@ func readFileFromArchive(t *testing.T, archive io.ReadCloser, name string, expec
+
+	logs := getDecompressLogs(t)
+	require.Len(t, logs, 1)
+	assert.Contains(t, logs[0].Message, "Decompressing with unpigz")
```
Please don't use log messages to test this.

sargun replied:
What else would you suggest doing?

stevvooe replied:
Not using log messages to test the functionality. I am not sure what these are testing exactly, so I don't have a good suggesting. Testing with log message hooks isn't great, as the log messages may change.

sargun replied:
I'm trying to test if pigz was used, and if it can be disabled. The alternative is to pass in some callback or something that's only used for testing which returns whether pigz was used or !pigz.

stevvooe commented:
> sort of optional in the registry and the registry to understand tar+compression.

No, the registry doesn't need to understand tar compression. It already supports transport compression via http. Disabling will be a change that breaks the cache but will be mostly transparent.

I am okay with this change as long as we only do this on decompression.

sargun commented:
@stevvooe Where in the registry does it support transport compression on upload?

tonistiigi commented:
> @tonistiigi had a note about the cmdstream but I leave that for him to comment 😄

I think the changes `cmdstream` in this PR actually address the comments made in 
https://github.com/moby/moby/pull/34218#issuecomment-317626581 . I recently wrote https://github.com/tonistiigi/copy/blob/go-copy/detect/detect.go#L85 for the same use-case. Maybe you find that one more readable but the behavior of the current implementation looks ok as well.

stevvooe commented:
> @stevvooe Where in the registry does it support transport compression on upload?

For downloading layers, a few modifications can be made to make this work or it can be done in a proxy. On upload, some changes would be required to support transfer encoding. Either way, compression of layers should be a transport concern and not part of their identity. It would be better if someone worked on that, rather than micro-optimizations further cementing the status quo.

sargun commented:
@stevvooe Maybe. It's something that I might be able to work on, but it's fairly complicated at first glance, at least how we use registry. Adding transfer encoding to the registry seems fairly straightforward, but the download side is more complicated. We serve files off of S3, and S3 doesn't support serving files with transparent compression. Proxying all of our downloads would be a significant cost, not only in CPU, but in network. It'd require a more complicated caching scheme, or some such.

stevvooe commented:
@sargun That is not really the only alternative. The goal is to have options for compression, rather than be pegged to _always_ using a single algorithm.

The issue here is that gzip is the limiting factor in pushing and pulling layers on fast data links. Currently, it is tied to the data format but compression is better used when it is tied to the media, be it compression for storage or compression for transfer. For example, inline compression might make sense for links under 100mbps because a CPU can keep up but it doesn't make any sense for 1Gbps links because the CPU can't compress fast enough and acts as a bottleneck in the transfer.

The continued issue here is that we compress layer artifacts whether or not it makes sense for the storage target or transfer medium. The mismatch here really shows itself when on a slow cpu (think raspberrypi) or on a really fast link.

Like I said, I think this PR is good, but it is only a small band aid on fixing the underlying problem.

sargun commented:
@stevvooe updated test PTAL.

stevvooe commented on pkg/archive/archive.go:
```diff
@@ -136,10 +138,37 @@ func DetectCompression(source []byte) Compression {
+}
+
+func gzDecompress(ctx context.Context, buf io.Reader) (io.ReadCloser, error) {
+	disablePigzEnv := os.Getenv("DISABLE_PIGZ")
```
Should this be `MOBY_DISABLE_PIGZ`?

stevvooe commented:
LGTM

@dmcgowan @tonistiigi PTAL

tonistiigi commented:
LGTM

AkihiroSuda commented on pkg/archive/archive.go:
```diff
@@ -136,10 +138,37 @@ func DetectCompression(source []byte) Compression {
+		}
+	}
+
+	unpigzPath, err := exec.LookPath("unpigz")
```
Please move this detection to `init()`, so as to reduce debug log lines?

AkihiroSuda commented on pkg/archive/archive_test.go:
```diff
@@ -1318,3 +1322,38 @@ func readFileFromArchive(t *testing.T, archive io.ReadCloser, name string, expec
+func TestDisablePigz(t *testing.T) {
+	_, err := exec.LookPath("unpigz")
+	if err != nil {
+		t.Log("Test will does not check full path when Pigz not installed")
```
typo: "will does not"

AkihiroSuda commented on pkg/archive/archive.go:
```diff
@@ -136,10 +138,37 @@ func DetectCompression(source []byte) Compression {
+}
+
+func gzDecompress(ctx context.Context, buf io.Reader) (io.ReadCloser, error) {
+	disablePigzEnv := os.Getenv("MOBY_DISABLE_PIGZ")
```
Can we have a documentation that lists up all supported environment variables?
Also, probably we could add them to `dockerd --help` (in a separate PR).

sargun replied:
I can add the documentation, RE: environment variables in a separate PR? Does that work?

thaJeztah commented:
re: documenting environment variables https://github.com/moby/moby/pull/35697#discussion_r161153846

There's some documentation in https://github.com/docker/cli/blob/master/docs/reference/commandline/cli.md#environment-variables and https://github.com/docker/cli/blob/master/docs/reference/commandline/dockerd.md, but looks like we need to clean that up; some may be only for the daemon, and others for both "cli" and "dockerd"

sargun commented:
Added here: https://github.com/docker/cli/pull/806

sargun commented:
@thaJeztah @AkihiroSuda good to go?

AkihiroSuda commented on pkg/archive/archive.go:
```diff
@@ -136,10 +149,34 @@ func DetectCompression(source []byte) Compression {
+		return gzip.NewReader(buf)
+	}
+
+	disablePigzEnv := os.Getenv("MOBY_DISABLE_PIGZ")
```
LGTM but you can also move this to init()

sargun replied:
That makes testing this more difficult.

AkihiroSuda commented:
@unclejack LGTY?

Commit:
Make image (layer) downloads faster by using pigz

The Golang built-in gzip library is serialized, and fairly slow
at decompressing. It also only decompresses on demand, versus
pipelining decompression.

This change switches to using the pigz external command
for gzip decompression, as opposed to using the built-in
golang one. This code is not vendored, but will be used
if it autodetected as part of the OS.

This also switches to using context, versus a manually
managed channel to manage cancellations, and synchronization.
There is a little bit of weirdness around manually having
to cancel in the error cases.

Signed-off-by: Sargun Dhillon <sargun@sargun.me>

sargun commented:
Rebased.

unclejack commented:
@sargun: Can you provide some benchmark numbers, please? It'd be nice to know how pigz performs vs gzip vs the Go gzip package from the standard library. I'm asking mostly because I've seen this paragraph in the pigz manual:
Decompression  can’t  be  parallelized,  at  least  not  without  specially  prepared  deflate  streams  for that purpose. As a result, pigz uses a single thread (the main thread) for decompression, but will create three other threads for reading, writing, and check calculation, which can speed up decompression under some circumstances. Parallel decompression can be turned off by specifying one process (-dp 1 or -tp 1).

sargun commented:
@unclejack I tested with the "legacy" code, gunzip, and unpigz with my test image. I had a local docker registry serving up artifacts, so the speed there was not part of the problem. The timings are below with each code path. These were the best two measurements of 5:

```
built-in pull:
2m36.376s
2m34.624s

pigz:
1m8.438s
1m7.825s

gunzip:
1m26.945s
1m26.026s
```

unclejack reviewed (approved):
@sargun: Thank you for posting the benchmark. This looks good.

This should provide a very nice speed boost for pulling images

LGTM

jboero commented:
Wow I thought nobody liked my pigz proposal but then I just noticed a RHEL 7 install failed with pigz as required prerequisite:

ec2-user@ip-172-31-9-105 ~> sudo yum install pigz
Loaded plugins: amazon-id, rhui-lb, search-disabled-repos
No package pigz available.
Error: Nothing to do

I love that pigz is now used but my PR had no packaging dependency and would fall back to regular gzip if pigz was unavailable.  D'oh!

Anybody in packaging please remove pigz from deps... RHEL 7 doesn't have a pigz in repos.

Thanks guys!

cpuguy83 commented:
No, there is no support for "weak depdendencies" until rather recently on
RPM, and not supported on RHEL.
Docker CE is for CentOS, which does have a pigz package.

On Mon, Apr 23, 2018 at 10:48 AM Sargun Dhillon <notifications@github.com>
wrote:

> At least for the Debian packaging, it just falls back.
>
> In Debian, they have the idea of “recommends” — does RHEL not?
>
> Sent from my iPhone
>
> On Apr 23, 2018, at 06:21, Brian Goff <notifications@github.com> wrote:
>
> @jboero <https://github.com/jboero> Agree having it as a requirement is
> not
> right, however there is no version of Docker for RHEL that has pigz support
> yet.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/moby/moby/pull/35697#issuecomment-383572276>, or mute
> the thread
> <
> https://github.com/notifications/unsubscribe-auth/AAtyRG7MtMu4keQ8oc4AyGeqUl_262oFks5trdVGgaJpZM4Q1g4j
> >
> .
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/moby/moby/pull/35697#issuecomment-383602273>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAwxZssdmd4oX3vzPvgNl1efMnjKiiUMks5trenPgaJpZM4Q1g4j>
> .
>

jboero commented:
Workaround - pigz is included in EPEL which works with RHEL (and CentOS)

lox commented:
It's super not clear how to actually make use of this, do we need to ensure that `unpigz` is available on hosts to see the speed up?

sargun commented:
@lox correct

GitHubDiom commented:
Hey, how do you guys make package decompression parallel?
I'd like to test parallel decompression but can't seem to do it.


dockerinfo
```shell
Client:
 Debug Mode: false

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 7
 Server Version: 19.03.13
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 8fba4e9a7d01810a393d5d25a3621dc101981175
 runc version: dc9208a3303feef5b3839f4323d9beb36df0a9dd
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 5.0.2-1.el7.elrepo.x86_64
 Operating System: CentOS Linux 7 (Core)
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 7.757GiB
 Name: master
 ID: KGRO:7J7A:7TKD:UNUO:ZQ7A:QYEJ:DJK4:OK4S:7JBS:U5IG:LBS3:2RJ7
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
```
- download is parallel
![image](https://user-images.githubusercontent.com/19339686/98667888-074edd00-238a-11eb-9bef-d7dbacbae5bf.png)
- But the extraction is serial.
![image](https://user-images.githubusercontent.com/19339686/98667921-133a9f00-238a-11eb-9bb0-cbfe9749b0fa.png)

thaJeztah commented:
@GitHubDiom this PR makes each individual "extract" run faster if `pigz` is installed, but layers still have to be extracted/applied in sequence; see https://github.com/moby/moby/issues/21814, https://github.com/moby/moby/issues/37957 for more details

GitHubDiom commented:
> @GitHubDiom this PR makes each individual "extract" run faster if `pigz` is installed, but layers still have to be extracted/applied in sequence; see #21814, #37957 for more details

Can docker pipeline the pull process for a specific layer?
As far as I know, three phases appear when pulling one layer, for example, download(Network-intensive) -> decompress (CPU-intensive) -> extract (IO-intensive).
For instance, extracting the layer may start immediately after the beginning of the layer image has been decompressed.

thaJeztah commented:
Theres some comments about decompress/extract in https://github.com/moby/moby/issues/21814#issuecomment-307454065 (and further)
